docs/interface.rst

```diff
@@ -325,74 +325,50 @@ There is a WebSocket-based status API located at ``/v1/status``.
-    - ``"conflicted"``: a boolean indicating if there is a conflict for this relative path
-  - ``"scanner"`` contains information about the local changes scanner
-    - ``"last-scan"``: ``null`` if no scan is completed yet, or the timestamp of the last completion
-  - ``"poller"`` contains information about the remote changes poller
-    - ``"last-poll"``: ``null`` if no scan is completed yet, or the timestamp of the last completion
-
-Clients should be tolerant of keys in the state they don't understand.
-Unknown state keys should be ignored.
-Note that ``"modified"`` is when the local state for thie item changed while ``"last-updated"`` is to do with the filesystem modification time.
-For example, a file may have an on-disk modification time that is older than the last time we updated our state about it, especially one that came from another device.
+...where the ``events`` list contains some non-zero number of event messages.
+The first message, upon connect, will likely contain many events: enough to give a consistent view of the current state.
+Thereafter, most messages will include only a single event (although clients should handle any number).
+
+Every event has a ``"kind"`` key describing what sort of message it is.
+Events will contain other keys; clients should be tolerant of keys in the state they don't understand.
 
 The client doesn't send any messages to the server; it is an error to do so.
+
+The follow event kinds are understood (see ``status.py`` for more details on the sending side, and ``cli.py`` for an example of receiving them):
+
+- ``"scan-completed"``: has key ``timestamp`` which is a unix-timestamp saying when we last looked for local changes.
+
+- ``"poll-completed"``: has a key ``timestamp`` describing when we last asked for remote changes.
+
+- ``"tahoe-connection-changed"``: describes the status of our connected Tahoe-LAFS client: ``connected`` and ``desired`` are the number of servers we are conencted to (and how many we want). Whether we are currently connected to enough is in a boolean ``happy``.
+
+- ``"error-occurred"``: An error, with ``folder`` (the name for the affected folder) ``timestamp`` and ``summary`` (human-readable string).
+
+- ``"folder-added"``: Key ``folder`` says which folder was added.
+
+- ``"folder-left"``: Key ``folder`` says which folder has gone away.
+
+- ``"upload-queued"``: some file (``relpath``) in a folder (``folder``) is queued for upload since ``queued-at``.
+
+- ``"upload-started"``: some file (``relpath``) in a folder (``folder``) has begun upload since``started-at``. An ``upload-queued`` event will always preceed this.
+
+- ``"upload-finished"``: a file (``relpath``) in a folder (``folder``) has completed. An ``upload-started`` will always preceed this.
+
+- ``"download-queued"``: same as upload version.
+
+- ``"download-started"``: same as upload version.
+
+- ``"download-finished"``: same as upload version.
+
+All timestamps are "seconds since the Unix epoch", as floats.
```

Comment:
This line is untrue. In the case of `error-occurred`, [`timestamp` is an integer](https://github.com/LeastAuthority/magic-folder/blob/d3cd024b7229475e0b5a63d05233e3837367b76a/src/magic_folder/status.py#L283), not a float. Consider either updating `error-occurred` to use a float too, or updating this line accordingly.

Further related to consistency: `timestamp` is used across three messages, but some other messages that effectively contain timestamps are calling those timestamps by other names (`queued-at`, `started-at`). Perhaps they should all just be `timestamp`? (Also, is there a reason why most of the messages have timestamps but some of them don't? Perhaps _every_ message should have a timestamp?)

Reply:
Note JSON has neither integers nor floats, [only numbers](https://www.rfc-editor.org/rfc/rfc8259#section-6).

Reply:
I thought I changed them all to `timestamp` so yes can change `queued-at` and `started-at` too then.

Should I just say "number" there? (I think I put "float" because that's more accurate for Python as to what JSON really does). Or should we serialize timestamps differently altogether (probably the other contender is ISO8601 strings?)

Note that the timestamps represent the time when the thing happened, not when the message is sent. But, sure, if timestamps can benefit other place we could add them. (It might be better to do that as a separate PR, as this was _supposed_ to be simply mechanical: going from "here is the state" to "here are events" -- but containing the same information).

Implementation-wise, internally we do already have and record timestamps for the things that include them, but not for everything. "folder add" doesn't record when it was done and there's nowhere sensible to record folder-delete timestamps? (unless we just do it ephemerally). Those are the only two right?

Reply:
> Should I just say "number" there? (I think I put "float" because that's more accurate for Python as to what JSON really does). Or should we serialize timestamps differently altogether (probably the other contender is ISO8601 strings?)

Since this is for the JSON interface, I think speaking in terms of JSON makes the most sense.  It also means that if a particular JSON implementation decides to spit out an `int` then the docs have not been proven wrong. :)

ISO8601 is _probably_ a reasonably good idea for timestamp representation but I don't know if it makes that much difference here.  It is more easily human-readable so it would make manual testing against the API or debugging by reading raw logs a bit more convenient.  Of course it would also be very easy to build a simple tool to improve the presentation of raw logs for any human who wants to get the info.

I'd probably lean towards putting ISO8601 timestamps on the todo list for v2 of the API and then forgetting about it for a while.

Reply:
Isn't `"foo": 0` still valid JSON, meaning "the float 0.0"...? (Similarly with `1` etc)

In any case, Python does indeed produce and consume JSON like that. (The RFC says "...which may be followed by a fraction part and/or an exponent part" so I think that's still standards-compliant).

Anyway, I'll change it to "number".